# Data Cleaning

### Imports

In [5]:
import pandas as pd
import numpy as nps

### Read Excel File Into DataFrame

In [6]:
df = pd.read_excel('Data/perfume_database.xlsx', usecols=['brand', 'perfume', 'launch_year', 'main_accords', 'notes'])

### Preview The DataFrame

In [7]:
df

brand                perfume  launch_year  \
0        18 21 Man Made  Sweet Tobacco Spirits       2016.0   
1      40 Notes Perfume          Cashmere Musk          NaN   
2      40 Notes Perfume     Exotic Ylang Ylang          NaN   
3      40 Notes Perfume        Exquisite Amber          NaN   
4      40 Notes Perfume           Oudwood Veil          NaN   
...                 ...                    ...          ...   
37921    Urban Rituelle       Lemongrass Blend          NaN   
37922    Urban Rituelle          Peach Blossom          NaN   
37923    Urban Rituelle            Pomegranate          NaN   
37924    Urban Rituelle                Vanilla          NaN   
37925    Urban Rituelle          Vanilla Blend          NaN   

                                            main_accords  \
0      ["wine", "vanilla", "sweet", "woody", "aromatic"]   
1      ["woody", "musky", "balsamic", "powdery", "war...   
2      ["yellow floral", "white floral", "sweet", "mu...   
3      ["balsamic", "vanilla", "amber", "musky", "war...   
4      ["oud", "amber", "fresh spicy", "balsamic", "w...   
...                                                  ...   
37921  ["citrus", "fresh spicy", "aromatic", "herbal"...   
37922  ["sweet", "honey", "white floral", "floral", "...   
37923  ["fruity", "citrus", "fresh", "woody", "sweet"...   
37924  ["vanilla", "caramel", "milky", "sweet", "powd...   
37925  ["aromatic", "fresh spicy", "vanilla", "floral...   

                                                   notes  
0      ["Citruses", "Saffron", "Tonka Bean", "Vanilla...  
1      ["Sandalwood", "Cedar", "White Musk", "Cashmer...  
2                    ["Ylang-Ylang", "Gardenia", "Musk"]  
3      ["Labdanum", "Styrax", "Benzoin", "Vanilla", "...  
4                         ["Kephalis", "Agarwood (Oud)"]  
...                                                  ...  
37921  ["Lemongrass", "Myrtle", "Grapefruit", "Eucaly...  
37922          ["Peach", "Honey", "Sweet Pea", "Mimosa"]  
37923         ["Pomegranate", "Citruses", "Red Berries"]  
37924                     ["Vanilla", "Caramel", "Milk"]  
37925                ["Vanilla", "Lavender", "Geranium"]  

[37926 rows x 5 columns]

* We are starting with 37,926 perfumes in our dataset!

### Cleaning The Text

In [8]:
#Not all the perfumes have top, middle, and base notes categorized, so, I am dropping that component. 
#All notes will still be accounted for in a perfume -- just not categorized by top/middle/base.

items_to_remove = [
    '[', ']', '"', '{', '}',
    'middle: ', 'top: ', 'base: '
]
def remove_items(text):
    for item in items_to_remove:
        text = text.replace(item, "")
    return text

In [35]:
df2 = pd.read_csv('small_cosine_sim_matrix.csv')
df2.columns[2]

'1'

### Cleaning up Notes Column

In [39]:
# string type 
df['notes'] = df['notes'].astype(str)

# lowercase
df['notes'] = df['notes'].str.lower()

# remove qoutes and brackets
df['notes'] = df['notes'].apply(remove_items)

### Cleaning up Main Accords Column

In [40]:
# string type 
df['main_accords'] = df['main_accords'].astype(str)

# lowercase
df['main_accords'] = df['main_accords'].str.lower()

# remove qoutes and brackets
df['main_accords'] = df['main_accords'].apply(remove_items)

In [9]:
df.head()

brand                perfume  launch_year  \
0    18 21 Man Made  Sweet Tobacco Spirits       2016.0   
1  40 Notes Perfume          Cashmere Musk          NaN   
2  40 Notes Perfume     Exotic Ylang Ylang          NaN   
3  40 Notes Perfume        Exquisite Amber          NaN   
4  40 Notes Perfume           Oudwood Veil          NaN   

                                        main_accords  \
0              wine, vanilla, sweet, woody, aromatic   
1  woody, musky, balsamic, powdery, warm spicy, a...   
2   yellow floral, white floral, sweet, musky, woody   
3        balsamic, vanilla, amber, musky, warm spicy   
4           oud, amber, fresh spicy, balsamic, woody   

                                               notes  
0  citruses, saffron, tonka bean, vanilla, exotic...  
1       sandalwood, cedar, white musk, cashmere wood  
2                        ylang-ylang, gardenia, musk  
3           labdanum, styrax, benzoin, vanilla, musk  
4                           kephalis, agarwood (oud)

### Cleaning up the Launch Year Column

In [10]:
df.isna().sum()

brand               0
perfume             3
launch_year     11210
main_accords        0
notes               0
dtype: int64

* Launch Year has over 11k missing values.
* I am NOT going to drop those rows because I want to retrain the data about their notes and accords! I will just fill in missing years with "0"
* However, I will drop rows missing any data on the accords, notes, and/or perfume name because that will affect my analysis the most

In [11]:
# Fill NaN values in 'launch_year' column with 0
df['launch_year'].fillna(0, inplace=True)

In [12]:
# Convert 'launch_year' column to integers
df['launch_year'] = df['launch_year'].astype(int)

In [13]:
df

brand                perfume  launch_year  \
0        18 21 Man Made  Sweet Tobacco Spirits         2016   
1      40 Notes Perfume          Cashmere Musk            0   
2      40 Notes Perfume     Exotic Ylang Ylang            0   
3      40 Notes Perfume        Exquisite Amber            0   
4      40 Notes Perfume           Oudwood Veil            0   
...                 ...                    ...          ...   
37921    Urban Rituelle       Lemongrass Blend            0   
37922    Urban Rituelle          Peach Blossom            0   
37923    Urban Rituelle            Pomegranate            0   
37924    Urban Rituelle                Vanilla            0   
37925    Urban Rituelle          Vanilla Blend            0   

                                            main_accords  \
0                  wine, vanilla, sweet, woody, aromatic   
1      woody, musky, balsamic, powdery, warm spicy, a...   
2       yellow floral, white floral, sweet, musky, woody   
3            balsamic, vanilla, amber, musky, warm spicy   
4               oud, amber, fresh spicy, balsamic, woody   
...                                                  ...   
37921       citrus, fresh spicy, aromatic, herbal, woody   
37922        sweet, honey, white floral, floral, powdery   
37923          fruity, citrus, fresh, woody, sweet, sour   
37924            vanilla, caramel, milky, sweet, powdery   
37925     aromatic, fresh spicy, vanilla, floral, herbal   

                                                   notes  
0      citruses, saffron, tonka bean, vanilla, exotic...  
1           sandalwood, cedar, white musk, cashmere wood  
2                            ylang-ylang, gardenia, musk  
3               labdanum, styrax, benzoin, vanilla, musk  
4                               kephalis, agarwood (oud)  
...                                                  ...  
37921         lemongrass, myrtle, grapefruit, eucalyptus  
37922                    peach, honey, sweet pea, mimosa  
37923                 pomegranate, citruses, red berries  
37924                             vanilla, caramel, milk  
37925                        vanilla, lavender, geranium  

[37926 rows x 5 columns]

### Cleaning up The Perfume Column

In [14]:
df.isna().sum()

brand           0
perfume         3
launch_year     0
main_accords    0
notes           0
dtype: int64

In [15]:
# Drop rows with missing values in the 'perfume' column
df.dropna(subset=['perfume'], inplace=True)

In [16]:
df.shape

(37923, 5)

In [17]:
df

brand                perfume  launch_year  \
0        18 21 Man Made  Sweet Tobacco Spirits         2016   
1      40 Notes Perfume          Cashmere Musk            0   
2      40 Notes Perfume     Exotic Ylang Ylang            0   
3      40 Notes Perfume        Exquisite Amber            0   
4      40 Notes Perfume           Oudwood Veil            0   
...                 ...                    ...          ...   
37921    Urban Rituelle       Lemongrass Blend            0   
37922    Urban Rituelle          Peach Blossom            0   
37923    Urban Rituelle            Pomegranate            0   
37924    Urban Rituelle                Vanilla            0   
37925    Urban Rituelle          Vanilla Blend            0   

                                            main_accords  \
0                  wine, vanilla, sweet, woody, aromatic   
1      woody, musky, balsamic, powdery, warm spicy, a...   
2       yellow floral, white floral, sweet, musky, woody   
3            balsamic, vanilla, amber, musky, warm spicy   
4               oud, amber, fresh spicy, balsamic, woody   
...                                                  ...   
37921       citrus, fresh spicy, aromatic, herbal, woody   
37922        sweet, honey, white floral, floral, powdery   
37923          fruity, citrus, fresh, woody, sweet, sour   
37924            vanilla, caramel, milky, sweet, powdery   
37925     aromatic, fresh spicy, vanilla, floral, herbal   

                                                   notes  
0      citruses, saffron, tonka bean, vanilla, exotic...  
1           sandalwood, cedar, white musk, cashmere wood  
2                            ylang-ylang, gardenia, musk  
3               labdanum, styrax, benzoin, vanilla, musk  
4                               kephalis, agarwood (oud)  
...                                                  ...  
37921         lemongrass, myrtle, grapefruit, eucalyptus  
37922                    peach, honey, sweet pea, mimosa  
37923                 pomegranate, citruses, red berries  
37924                             vanilla, caramel, milk  
37925                        vanilla, lavender, geranium  

[37923 rows x 5 columns]

### Checking for null string values such as "nan"

In [18]:
df['main_accords'].value_counts()

main_accords
nan                                                      969
rose, floral, citrus                                      85
floral                                                    81
white floral, animalic, floral                            80
oud, fresh spicy, balsamic, woody                         61
                                                        ... 
woody, citrus, balsamic, aromatic, musky, powdery          1
sweet, caramel, warm spicy, balsamic, floral, powdery      1
aromatic, white floral, musky, patchouli, fresh spicy      1
floral, fresh, rose, patchouli, fresh spicy                1
vanilla, caramel, milky, sweet, powdery                    1
Name: count, Length: 33069, dtype: int64

In [19]:
df['notes'].value_counts()

notes
nan                                                                                   957
floral notes                                                                           66
rose                                                                                   59
agarwood (oud)                                                                         58
lavender                                                                               55
                                                                                     ... 
green notes, floral notes, spicy notes                                                  1
oriental flower notes, spicy notes, green accord                                        1
bergamot, woodsy notes, oak moss, jasmine                                               1
lavender, sage, basil, vetiver, sandalwood, tonka bean, orange, lemon, green notes      1
vanilla, lavender, geranium                                                             1
Name

### I am going to drop the rows containing "nan" because that is not a valid fragrance note or accord.

In [20]:
# Drop rows where 'main_accords' or 'notes' column contains "nan"
df_filtered = df[(df['main_accords'].str.lower() != 'nan') & (df['notes'].str.lower() != 'nan')]

# Reset index
df_filtered.reset_index(drop=True, inplace=True)

### How many rows were dropped so far?
* We started with 37,926 rows and now have 36,954.
* Therefore 972 rows were dropped!

In [21]:
df_filtered.shape

(36954, 5)

In [22]:
df_filtered

brand                perfume  launch_year  \
0        18 21 Man Made  Sweet Tobacco Spirits         2016   
1      40 Notes Perfume          Cashmere Musk            0   
2      40 Notes Perfume     Exotic Ylang Ylang            0   
3      40 Notes Perfume        Exquisite Amber            0   
4      40 Notes Perfume           Oudwood Veil            0   
...                 ...                    ...          ...   
36949    Urban Rituelle       Lemongrass Blend            0   
36950    Urban Rituelle          Peach Blossom            0   
36951    Urban Rituelle            Pomegranate            0   
36952    Urban Rituelle                Vanilla            0   
36953    Urban Rituelle          Vanilla Blend            0   

                                            main_accords  \
0                  wine, vanilla, sweet, woody, aromatic   
1      woody, musky, balsamic, powdery, warm spicy, a...   
2       yellow floral, white floral, sweet, musky, woody   
3            balsamic, vanilla, amber, musky, warm spicy   
4               oud, amber, fresh spicy, balsamic, woody   
...                                                  ...   
36949       citrus, fresh spicy, aromatic, herbal, woody   
36950        sweet, honey, white floral, floral, powdery   
36951          fruity, citrus, fresh, woody, sweet, sour   
36952            vanilla, caramel, milky, sweet, powdery   
36953     aromatic, fresh spicy, vanilla, floral, herbal   

                                                   notes  
0      citruses, saffron, tonka bean, vanilla, exotic...  
1           sandalwood, cedar, white musk, cashmere wood  
2                            ylang-ylang, gardenia, musk  
3               labdanum, styrax, benzoin, vanilla, musk  
4                               kephalis, agarwood (oud)  
...                                                  ...  
36949         lemongrass, myrtle, grapefruit, eucalyptus  
36950                    peach, honey, sweet pea, mimosa  
36951                 pomegranate, citruses, red berries  
36952                             vanilla, caramel, milk  
36953                        vanilla, lavender, geranium  

[36954 rows x 5 columns]

### Dropping duplicates
* There are 11 duplicates!

In [23]:
df_filtered[df_filtered.duplicated()]

brand    perfume  launch_year  \
2889           Commodity      Paper         2013   
3557   Demeter Fragrance   Geranium            0   
3558   Demeter Fragrance   Lavender            0   
3561   Demeter Fragrance  Patchouli            0   
17368           Molinard  Patchouli         2015   
18688          Prouvenco    Lavande            0   
31108          Blend Oud     Meksar         2014   
31109          Blend Oud  Rouh Aoud            0   
31110          Blend Oud      Sahar            0   
31111          Blend Oud       Bark            0   
31112          Blend Oud     Ghazal            0   

                                            main_accords  \
2889              woody, amber, musky, animalic, powdery   
3557   fresh spicy, aromatic, herbal, warm spicy, floral   
3558        aromatic, fresh spicy, floral, herbal, sweet   
3561      patchouli, warm spicy, earthy, balsamic, woody   
17368     patchouli, warm spicy, earthy, balsamic, woody   
18688       aromatic, fresh spicy, floral, herbal, sweet   
31108     woody, musky, amber, balsamic, earthy, powdery   
31109  woody, warm spicy, balsamic, fresh spicy, rose...   
31110      fresh spicy, woody, balsamic, aromatic, musky   
31111  citrus, warm spicy, cinnamon, woody, amber, le...   
31112        floral, citrus, sweet, fruity, white floral   

                                                   notes  
2889              amber, woody notes, cedar, iso e super  
3557                                            geranium  
3558                                            lavender  
3561                                           patchouli  
17368                                          patchouli  
18688                                           lavender  
31108  cedar, angelica, rose, orris, guaiac wood, amb...  
31109  turkish rose, damask rose, rose de mai, agarwo...  
31110  agarwood (oud), styrax, incense, papyrus, cash...  
31111  rose, cinnamon, spicy notes, leather, patchoul...  
31112  lily-of-the-valley, peony, freesia, jasmine, r...

In [24]:
# Remove duplicate rows
df_filtered.drop_duplicates(inplace=True)

/var/folders/x_/qq_k4mdj52g85pqnx0_zznsw0000gn/T/ipykernel_2605/4168984718.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df_filtered.drop_duplicates(inplace=True)


### Let's get a final look at our clean dataframe

In [25]:
df_filtered

brand                perfume  launch_year  \
0        18 21 Man Made  Sweet Tobacco Spirits         2016   
1      40 Notes Perfume          Cashmere Musk            0   
2      40 Notes Perfume     Exotic Ylang Ylang            0   
3      40 Notes Perfume        Exquisite Amber            0   
4      40 Notes Perfume           Oudwood Veil            0   
...                 ...                    ...          ...   
36949    Urban Rituelle       Lemongrass Blend            0   
36950    Urban Rituelle          Peach Blossom            0   
36951    Urban Rituelle            Pomegranate            0   
36952    Urban Rituelle                Vanilla            0   
36953    Urban Rituelle          Vanilla Blend            0   

                                            main_accords  \
0                  wine, vanilla, sweet, woody, aromatic   
1      woody, musky, balsamic, powdery, warm spicy, a...   
2       yellow floral, white floral, sweet, musky, woody   
3            balsamic, vanilla, amber, musky, warm spicy   
4               oud, amber, fresh spicy, balsamic, woody   
...                                                  ...   
36949       citrus, fresh spicy, aromatic, herbal, woody   
36950        sweet, honey, white floral, floral, powdery   
36951          fruity, citrus, fresh, woody, sweet, sour   
36952            vanilla, caramel, milky, sweet, powdery   
36953     aromatic, fresh spicy, vanilla, floral, herbal   

                                                   notes  
0      citruses, saffron, tonka bean, vanilla, exotic...  
1           sandalwood, cedar, white musk, cashmere wood  
2                            ylang-ylang, gardenia, musk  
3               labdanum, styrax, benzoin, vanilla, musk  
4                               kephalis, agarwood (oud)  
...                                                  ...  
36949         lemongrass, myrtle, grapefruit, eucalyptus  
36950                    peach, honey, sweet pea, mimosa  
36951                 pomegranate, citruses, red berries  
36952                             vanilla, caramel, milk  
36953                        vanilla, lavender, geranium  

[36943 rows x 5 columns]

In [26]:
df_filtered.shape

(36943, 5)

### Looks good! We started out with 37,926 rows and now we have 36,943. 
* I dropped 983 rows in total 

In [27]:
# Save the DataFrame to a CSV file in the current directory
df_filtered.to_csv('clean_perfume_data.csv', index=False)

### Creating a small subset of data for running on streamlit
* For my EDA, I will use the full size dataset 'clean_perfume_data.csv'

In [38]:
# Sample 5,000 random rows
fivek_subset_df = df.sample(n=5000, random_state=1)

# Save the subset to a new file
fivek_subset_df.to_csv('fivek_subset_data.csv', index=False)